gateway-service-knoxtoken/src/test/java/org/apache/knox/gateway/service/knoxtoken/JWKSResourceTest.java

```diff
@@ -0,0 +1,144 @@
+    kpg.initialize(2048);
+    KeyPair KPair = kpg.generateKeyPair();
+
+    publicKey = (RSAPublicKey) KPair.getPublic();
+    privateKey = (RSAPrivateKey) KPair.getPrivate();
+  }
+
+  private void init() throws KeystoreServiceException, KeyStoreException {
+    final KeystoreService ks = EasyMock.createNiceMock(KeystoreService.class);
+    services = EasyMock.createNiceMock(GatewayServices.class);
+    context = EasyMock.createNiceMock(ServletContext.class);
+    request = EasyMock.createNiceMock(HttpServletRequest.class);
+
+    jwksResource = EasyMock.partialMockBuilder(JWKSResource.class)
+        .addMockedMethod("getPublicKey", String.class).createMock();
+
+    EasyMock.expect(services.getService(ServiceType.KEYSTORE_SERVICE))
+        .andReturn(ks).anyTimes();
+    EasyMock.expect(
+        context.getAttribute(GatewayServices.GATEWAY_SERVICES_ATTRIBUTE))
+        .andReturn(services).anyTimes();
+    EasyMock.expect(jwksResource.getPublicKey(null)).andReturn(publicKey)
+        .anyTimes();
+    EasyMock.replay(jwksResource, context, request, services, ks);
+  }
+
+  @Test
+  public void testJWKSrequest()
+      throws KeystoreServiceException, KeyStoreException {
+    init();
+    Response retResponse = jwksResource.getJwksResponse();
+    Assert.assertEquals(Response.Status.OK.getStatusCode(),
+        retResponse.getStatus());
+  }
+
+  /**
+   * End to End test that verifies the token acquired from JWKS endpoint.
+   *
+   * @throws KeystoreServiceException
+   * @throws KeyStoreException
+   * @throws ParseException
+   * @throws JOSEException
+   */
+  @Test
+  public void testE2E()
+      throws KeystoreServiceException, KeyStoreException, ParseException,
+      JOSEException {
+    init();
+    /* get a signed JWT token */
+    final JWT testToken = getTestToken();
+    /* get JWKS keyset */
+    final Response retResponse = jwksResource.getJwksResponse();
+
+    /* following lines just verifies the token */
+    final JWKSet jwks = JWKSet.parse(retResponse.getEntity().toString());
+    Assert.assertTrue("No keys found", jwks.getKeys().size() > 0);
+    final JWK jwk = jwks.getKeys().get(0);
+    Assert.assertNotNull("No private key found", jwk.toRSAKey().toPublicKey());
+    final PublicKey pk = jwk.toRSAKey().toPublicKey();
+    final JWSVerifier verifier = new RSASSAVerifier((RSAPublicKey) pk);
+    Assert.assertTrue("Cannot verify the token, wrong certificate",
+        testToken.verify(verifier));
+  }
+
+  private JWT getTestToken() {
+    String[] claimArray = new String[4];
+    claimArray[0] = "KNOXSSO";
+    claimArray[1] = "joe@example.com";
+    claimArray[2] = null;
+    claimArray[3] = null;
+
+    final String algorithm = "RS256";
+    final JWT token = new JWTToken(algorithm, claimArray,
+        Collections.singletonList("aud"), false);
+    final JWSSigner signer = new RSASSASigner(privateKey, true);
+    token.sign(signer);
+    return token;
+  }
+
```

Comment:
Can you add a test for when shared secret is being used to insure that we do not NPE or the like somewhere along the line? Just to tease out assumptions.

Reply:
Right, yes I will.